### ISSUE TITLE
`prefer-default-parameters` does not properly consider nulls

### ISSUE THREAD

**Issue by OlsonDev** [sindresorhus/eslint-plugin-unicorn#2422]
The rule `prefer-default-parameters` wants this:

```js
function doSomething(couldBeNull) {
  const normalized = couldBeNull ?? ''
  // Do something with normalized
}
```

to look like what exactly?
```js
function doSomething(couldBeNull = '') {
  // Do something with couldBeNull, but how to handle null without violating rule?
}
```

This is a problem because:
```js
doSomething(someExternalLibraryIHaveNoControlOverThatCouldReturnNull())
```

Of course I could fix all the call sites since I own `doSomething`:
```js
doSomething(someExternalLibraryIHaveNoControlOverThatCouldReturnNull() ?? '')
```

But, no thank you.

**Comment by sindresorhus** [comment]
The rule doesn't know the type of the parameter, so it cannot do anything about it. You can disable the rule if you prefer `??` instead.

We can document the caveat.

**Comment by OlsonDev** [comment]
My point is it's not a preference -- if I use the rule, there's nothing I can do to my code to satisfy the rule and have correct code.

**Comment by fregante** [comment]
Yeah the difference between default parameters and `??` is quite annoying. Essentially it makes `prefer-default-parameters` "impossible".

Unicorn essentially disregards/[forbids `null`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-null.md) as a value, so if you do encounter it you're kind of expected to eliminate it.

In this case for example it argues you should not accept `null` in your function, so if the user does, you could/should throw a `TypeError`. The user (or you) would then have to use

```js
doSomething(someExternalLibraryIHaveNoControlOverThatCouldReturnNull() ?? undefined)
```

🤷‍♂️ 😅 

If you don't like/want that, there's no alternative to disabling the rule.

**Comment by OlsonDev** [comment]
Just documenting it feels like a band-aid to me. Unicorn is entitled to be opinionated -- it's just unfortunately because most of these rules are great in isolation, but a few of them are significantly impacted by the opinion about `null`. It would be nicer if this particular rule could support a relaxed mode or have some "edge" cases handled nicer.

Some native methods return `null`, too.

![image](https://github.com/user-attachments/assets/0fffe3bb-2fc0-435a-8025-1af0965bd8ba)

So, in this case my code would have to change from:
```javascript
processMatches(userInputString.match(/bar/))
```
to:
```javascript
processMatches(userProvidedString.match(/bar/) ?? undefined)
```

Which, of course, is awkward/verbose.

**Comment by fregante** [comment]
> if this particular rule could support a relaxed mode or have some "edge" cases handled.

I don't see how, the rule applies to function declarations, not to call sites. Either you accept null or you don't. This is why I say the rule is incompatible with the existence of null.

**Comment by OlsonDev** [comment]
Maybe something like (with a configuration of this rule influencing change in behavior):

Fail:
```javascript
function abc(foo) {
  foo = foo || 'bar'
}

function abc(foo) {
  foo = foo === undefined ? 'bar' : foo
}
```

Pass:
```javascript
function abc(foo) {
  foo = foo ?? 'bar'
}

function abc(foo) {
  foo = foo == null ? 'bar' : foo
}

function abc(foo) {
  foo = foo === null ? 'bar' : foo
}
```

Thinking about the example in the docs more, it's a little weird to me that this fails:
```javascript
function abc(foo) {
	const bar = foo || 'bar';
}
```

That is, you might want _another_ variable that differs based on what was passed in. That is, you're not mutating the argument itself. 🤷🏻‍♂️

**Comment by fregante** [comment]
None of those use default parameters. The rule is called `prefer-default-parameters`

**Comment by fregante** [comment]
> ``` js
> function abc(foo) {
>   foo = foo || 'bar'
> }
> ``` 

This suggestion makes no sense because not only it still alters how it deals with `null`, but *also* other falsy values.

> ``` js
> function abc(foo) {
>   foo = foo === undefined ? 'bar' : foo
> }
> ``` 

This is just about the only code that can be replaced "safely", but also it's not something I've seen very often.

@sindresorhus @fisker should the rule be (optionally?) reduced to only handle this scenario?
